src/selectors/modular-pipelines.js

```diff
@@ -1,34 +1,220 @@
+  (modularPipelineIDs, modularPipelineEnabled) => ({
+    total: modularPipelineIDs.length,
+    enabled: modularPipelineIDs.filter((id) => modularPipelineEnabled[id])
+      .length,
+  })
+);
+
+/**
+ * List all the child nodes in each modular pipeline
+ */
+export const getModularPipelineChildren = createSelector(
+  [getModularPipelineIDs, getNodeIDs, getNodeModularPipelines],
+  (modularPipelineIDs, nodeIDs, nodeModularPipelines) => {
+    const modPipNodes = arrayToObject(modularPipelineIDs, () => ({}));
+    nodeIDs.forEach((nodeID) => {
+      nodeModularPipelines[nodeID]?.forEach((modPipID) => {
+        if (!modPipNodes[modPipID]) {
+          modPipNodes[modPipID] = {};
+        }
+        modPipNodes[modPipID][nodeID] = true;
+      });
+    });
+    return modPipNodes;
+  }
+);
+
+/**
+ * List all the parent modular pipelines for a given modular pipeline
+ */
+export const getModularPipelineParents = createSelector(
+  [getModularPipelineIDs],
+  (modularPipelineIDs) =>
+    arrayToObject(modularPipelineIDs, (modPipID) =>
+      modPipID.split('.').reduce((parents, part, i, parts) => {
+        if (i < parts.length - 1) {
+          let parent = part;
+          if (i > 0) {
+            parent = parents[parents.length - 1] + '.' + parent;
+          }
+          parents.push(parent);
+        }
+        return parents;
+      }, [])
+    )
+);
+
+/**
+ * Set disabled status if the node is specifically hidden, and/or via a tag/view/type/modularPipeline
+ */
+export const getModularPipelineParentsContracted = createSelector(
+  [
+    getModularPipelineIDs,
+    getModularPipelineParents,
+    getModularPipelineContracted,
+  ],
+  (modularPipelineIDs, modularPipelineParents, modularPipelineContracted) =>
+    arrayToObject(modularPipelineIDs, (modPipID) =>
+      modularPipelineParents[modPipID].some(
+        (id) => modularPipelineContracted[id]
+      )
+    )
+);
+
+/**
+ * Get a list of input/output edges of visible modular pipeline pseudo-nodes
+ * by examining the edges of their childen, and also create transitive edges
+ * between collapsed modular pipelines
+ */
+export const getModularPipelineEdges = createSelector(
+  [
+    getModularPipelineIDs,
+    getModularPipelineChildren,
+    getNodeModularPipelines,
+    getEdgeIDs,
+    getEdgeSources,
+    getEdgeTargets,
+  ],
+  (
+    modularPipelineIDs,
+    modularPipelineChildren,
+    nodeModularPipelines,
+    edgeIDs,
+    edgeSources,
+    edgeTargets
+  ) => {
+    // List of new edges generated from modular pipelines:
+    const edges = {
+      ids: {},
+      sources: {},
+      targets: {},
+    };
+
+    const addNewEdge = (source, target) => {
+      const id = [source, target].join('|');
+      edges.ids[id] = true;
+      edges.sources[id] = source;
+      edges.targets[id] = target;
+    };
+    modularPipelineIDs.forEach((modPipID) => {
+      const modPipNodes = modularPipelineChildren[modPipID];
+      edgeIDs.forEach((edgeID) => {
```

Comment:
Wondering if this could get slow with lots of edges with the nested loops, maybe it could be made linear looking at it another way if so?

Reply:
Yeah this is a concern of mine also - but I'm honestly not sure how to do it differently? I'll look into it. The good news is that this will only run on first loading the data, and won't re-run when toggling nodes'/MPs' disabled state.